testing-styleguide/README.md

```diff
@@ -0,0 +1,496 @@
+
+- That the inner layer (manager) is called correctly, with the expected set of parameters
+- That the input is correctly handled (e.g. that an ActiveRecord attribute is correctly updated at the database level)
+- That the response is the expected (e.g. response.body, response.status, etc)
+
+Note here we usually hit a testing database with a set of transactions, and assert changes in the ActiveRecord itself. In addition to that, external APIs can be mocked with their expect behavior for a given scenario.
+
+Let's use a Controller that handles a Hiring update as example. It gets a hash with the updated hiring values and let's imagine we are interested in the response status code being 200 if all goes well with the updated hiring hash in a json format. We also expect our controller to call a manager in order to perform the task in hand: here we name it `HiringManagerUpdater.update`.
+
+❌ Bad
+
+```ruby
+# !! Do not replace the manager with a fake implementation
+let(:hiring_manager_updater) { instance_double ::Services::Companies::Hirings::HiringManagerUpdater }
+
+context 'when controller is called with valid params' do
+  before do
+    put 'companies/hiring', params: params
+  end
+
+  it "should call the hiring manager updater with appropriate arguments" do
+    # !! Do not assert interactions with the stubbed manager
+    expect(hiring_manager_updater).to have_received(:update).with(expected_params)
+  end
+end
+```
+
+✅ Good
+
+```ruby
+context 'when controller is called with valid params' do
+  # Use FactoryBot.create to create an instance of your data object and persist it to the testing database
+  let(:hiring) { create(:hiring, status: Hiring.status.waiting) }
+  let(:params) do
+    {
+      hiring: {
+        id: hiring.id,
+        status: Hiring.status.approved
+      }
+    }
+  end
+  # Describe the expected set of (transformed) parameters your manager needs to properly work
+  let(:manager_params) do
+    ActionController::Parameters.new( params.map { |k,v| [k, v.to_s] }.to_h )
+      .require(:hiring).permit(:id, :status)
+  end
+  let(:update_hiring) do
+    put 'companies/hiring', params: params
+  end
+
+  before do
+    # Only allow calls to HiringManagerUpdater.create that happen with the expected set of parameters
+    allow_any_instance_of(::Services::Companies::Hirings::HiringManagerUpdater).to receive(:update).with(manager_params).and_call_original
+    # If you don't expect exceptions to be rescued in this context, add this line
+    allow(::Services::Logger::ExceptionLoggerService).to receive(:call).and_raise('Oups. No exception should have been seen!')
+    update_hiring
```

Comment:
@joaophellip If well implemented, the Entrypoint Layer will raise any error that the Manager has returned at 'callback.on_failed' 

Enriching the discussion a bit, I believe that we don't need to assert if some error has not been raised. It seems redundant to me because if we are already asserting, for example, whether hiring was updated, we are also asserting that no errors were raised, because, if so, the hiring would not be updated.

It makes more sense to me only to assert expected errors to be raised (e.g. friendly errors messages), also preventing us from hiding unexpected errors from our tests that should break them